add skill states and categories and enable barge in

ken-mycroft opened the pull request:
## Description
JIRA SKILL-224 and associated

## How to test
Will require Alarm and News skill PRs as well (same name as this one) 
Set alarm for 1 minute from now, play the news, say stop, only the alarm should stop, the news should continue to play

## Contributor license agreement signed?
yes

chrisveilleux commented on mycroft/skills/common_play_skill.py:
```diff
@@ -19,6 +19,7 @@
 from .mycroft_skill import MycroftSkill
 from .audioservice import AudioService
 
+from mycroft.util.log import LOG
```
this should be placed above the local relative imports

ken-mycroft replied:
done

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -17,13 +17,15 @@
     AdaptService, AdaptIntent, FallbackService, PadatiousService, IntentMatch
 )
 from .intent_service_interface import open_intent_envelope
+from mycroft.messagebus.message import Message
```
move up to where the other mycroft imports are

ken-mycroft replied:
done

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -103,12 +106,25 @@ def __init__(self, bus):
         def add_active_skill_handler(message):
-            self.add_active_skill(message.data['skill_id'])
+            category = 'undefined'
+            if message.data.get("skill_cat", None) is not None:
```
None is the default value returned by get() if no match is found.  you don't need to specify it.

ken-mycroft replied:
done

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -103,12 +106,25 @@ def __init__(self, bus):
+        self.bus.on('deactivate_skill_request', remove_active_skill_handler)
         self.active_skills = []  # [skill_id , timestamp]
+
+        # BUG fix this!
```
Could you elaborate, in case someone else decides to fix it?

ken-mycroft replied:
Its actually not a bug its just a poor design. i removed the comment

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -166,6 +187,7 @@ def do_converse(self, utterances, skill_id, lang, message):
             lang (str): current language
             message (Message): message containing interaction info.
         """
+        LOG.debug("do_converse()-utterances:%s, active_skills=%s" % (utterances, self.active_skills))
```
%s notation for string variable resolution is old-school.  use an f-string or {}.format instead.  this happens in a few places in this PR

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -197,11 +219,11 @@ def remove_active_skill(self, skill_id):
             skill_id (str): skill to remove
         """
-        for skill in self.active_skills:
+        for skill in copy(self.active_skills):
```
if you cast this into a tuple instead of copying it, the intent of having an immutable copy for the loop would be more evident, IMO

forslund replied:
I disagree here (as per usual with me and Chris ;) ), an explicit copy is more expressive of the intention than an implicit copy as part of a cast.

ken-mycroft replied:
I love it. Before I started here I would take code review comments seriously and get upset on occasion but if this job has taught me anything it has taught me to be less thin skinned and more dismissive of stylistic input. FYI guys, my comment on this would have been "why is this not a lock? Casting and copy are not thread safe and have issues with the GIL"? Substance before style.

forslund replied:
I think the copy here is to handle the modification of the list below, not for thread safety, I'm sure a lock would be an improvement.

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -296,14 +321,30 @@ def handle_utterance(self, message):
+
+                        self.add_active_skill(match.skill_id, 
+                                self.skill_categories[match.skill_id])
+                        LOG.error("IntentService:handle_utterance() - Stop Msg Sent, Adding old style skill (id=%s) to active_skills array=%s" % (match.skill_id, self.active_skills))
```
consider changing this to a warning, kind of like a deprecation warning.  nothing is necessarily wrong.  in fact not all skills necessarily need to implement state, as we have discussed.

ken-mycroft replied:
My mistake, it is actually a debug

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -296,14 +321,30 @@ def handle_utterance(self, message):
 
                 # Launch skill if not handled by the match function
                 if match.intent_type:
+                    LOG.error("IntentService:handle_utterance() - Matched intent type! type:%s, \ndata:%s" % (match.intent_type, match.intent_data))
```
Why is this an error?

ken-mycroft replied:
Its not - my mistake i left some extra logging in

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -332,12 +373,23 @@ def _converse(self, utterances, lang, message):
 
+
+        # first check for system levl skills
+        tmp_active_skills = copy(self.active_skills)
```
again, casting to a tuple may make more sense than making a copy

ken-mycroft replied:
And again, this should really have a lock around it

chrisveilleux commented on mycroft/skills/intent_service.py:
```diff
@@ -332,12 +373,23 @@ def _converse(self, utterances, lang, message):
+        # first check for system levl skills
+        tmp_active_skills = copy(self.active_skills)
+        for skill in tmp_active_skills:
+            if skill[2] == 'system':
```
what are skill[2] and skill[0].  improve readability by adding a statement that names the values in these indices.

forslund replied:
or using NamedTuple instead of a regular tuple for the active_skills

ken-mycroft replied:
This also brings up a really good issue regarding how the PR process is handled and I am interested in input from both of you regarding this concept. 

BTW I agree a named tuple here would be most appropriate and I also agree additional documentation would be beneficial, however, you do realize this code pre-dates me. 

The use of "skill[some magic number] " is something which was here before I touched this code. I simply added an additional value to this list entry but now it seems like I own it and should go back through the code and clean up this old code because I am submitting a PR against it. 

Where does that stop. Suppose I do create a named tuple and refactor any references to skill[] to use that. Now in so doing I may touch another file which also has some questionable code in it. Should I also be required to correct that as a part of this PR process? What if that leads me to yet another file and so on. Also of concern is the fact that I work here. I get paid by the hour so I can take the time to respond to every comment against my PRs, I can address other people's opinions about how I decided to architect something, but contributors do not have that luxury. For most members of our community this is not their full time job and they don't get paid for this effort.

In other words, regarding PRs and our code review process, what is the objective here and what are the trade-offs? Do we want programmers to refactor the code with each PR at the expense of time or do we want to get more code in at the expense of leaving old ugly code ugly. 

I would be interested in both your thoughts.

forslund replied:
I'm sure I would have reacted to this other than to point out that NamedTuples are a nice alternative to variables for 0 and 1 :) (I'm a bit in love with them) since they work just as normal tuples so no need to change everywhere at once.

To answer your question, I try to keep PR's focused to as small an area as possible, it makes. If I get the opportunity to clean up something related to my update (notice a bad docstring, log message, small structure that can be re-ordered) , I do it in a separate commit. Never cross the beams! (refactoring & feature)

In short, you're not required to refactor / take ownership on any code you basically only move around or extend.

ken-mycroft replied:
What's interesting is I think Jarbas alerted me to the issue that changing this to a named tuple would be a breaking change.

forslund replied:
Ok, I can't see it but maybe I'm wrong, it wouldn't be the first time and won't be the last time.

ken-mycroft replied:
Existing old code refers to these values using the skill[offset] style but if I change them to a named tuple won't the old code have to be changed to refer to the entries at offset by name? Can you still refer to the individual elements of a named tuple by offset (int) and pass the linter? I could be wrong here.

ken-mycroft replied:
I'm going to leave this alone for now but I agree with the sentiment. Note the structure is documented in the existing comment around line 124 where the active_skills array is initialized. I added the category to this definition for clarity.

forslund replied:
Yes named tuples are great that way. both indexes and names work:
```python
>>> MyTup = namedtuple('MyTup', ('alpha', 'beta', 'gamma'))
>>> t = MyTup(1, 2, 3)
>>> t[1]
2
>>> t.beta
2
>>>
```

Leaving alone sounds good.

chrisveilleux commented on mycroft/skills/mycroft_skill/mycroft_skill.py:
```diff
@@ -165,6 +166,38 @@ def __init__(self, name=None, bus=None, use_settings=True):
+            return
+
+        if new_state not in self.skill_control.states:
+            self.log.warning("invalid state change, from %s to %s" % (self.skill_control.state, new_state))
```
this is a bit misleading.  the state change itself is not valid, the value of the new state is.  the log message doesn't say which the invalid value is, the old or the new.

ken-mycroft replied:
No, this is warning you the new state is NOT a valid value as it does not exist in your enumeration of states object (self.skill_control.states).

chrisveilleux commented on mycroft/skills/mycroft_skill/mycroft_skill.py:
```diff
@@ -165,6 +166,38 @@ def __init__(self, name=None, bus=None, use_settings=True):
 
+        self.skill_control = SkillControl()
+
+    def change_state(self, new_state):
```
add a docstring

ken-mycroft replied:
done

chrisveilleux commented on mycroft/skills/mycroft_skill/mycroft_skill.py:
```diff
@@ -673,13 +706,9 @@ def send_email(self, title, body):
-        """
-        self.bus.emit(Message('active_skill_request',
-                              {'skill_id': self.skill_id}))
+        if self.skill_control.category == 'undefined':
```
add a docstring

ken-mycroft replied:
For what? There is one in skill_control. I am confused.

forslund replied:
I think he means that the `make_active()` method should have a docstring. In this case it's extra useful since it's one of the methods in the "skill api" which is automatically documented and published on readthedocs: https://mycroft-core.readthedocs.io/en/latest/source/mycroft.html#mycroft.MycroftSkill.make_active

ken-mycroft replied:
Oh I see. Well make_active() is nothing new. It was around before this PR and to be honest I am not sure of its value. This kind of relates to the earlier comment which is how much code do I inherit when I add a PR :-)_

forslund replied:
No make_active is not a new thing but the PR removes it's docstring...

I guess this is another part of the issue, keeping skill compatibility...There are skills that relies on the functionality of the public methods. Ideally the skill creators that use this method shouldn't be affected even if you extend it's functionality, like you have here.

I think the docstring should remain, but be updated to indicate that this is deprecated method and the new way for keeping skills in converse priority is...

ken-mycroft replied:
Oh, my mistake. I did not mean to remove that method's doc string nor to deprecate it. In fact I believe it still works as always for skills of the 'undefined' category. I was just confused as to what the ask was. Now I understand. I will put back the doc string back in.

ken-mycroft replied:
Doc string added and doc string put back.

chrisveilleux commented on mycroft/skills/mycroft_skill/skill_control.py:
```diff
@@ -0,0 +1,55 @@
+# limitations under the License.
+#
+class SkillControl():
+    """
```
Why is this only like 40 characters wide?

ken-mycroft replied:
No good reason.

chrisveilleux commented on mycroft/skills/mycroft_skill/skill_control.py:
```diff
@@ -0,0 +1,55 @@
+    state = 'inactive'
+    states = None
+    category = 'undefined'
+
```
perhaps the state change method in MycroftSkill belongs in this class?

ken-mycroft replied:
No, I like the change state method in the skill base class. Yes the names StateControl and/or StateManager would also be good, however, I chose SkillControl.

chrisveilleux reviewed (changes requested):
Just a few minor things to adjust.

JarbasAl commented:
I don't agree with this PR implementation, i think it misunderstands the way the converse system works as illustrated by the #TODO fix bug comment.

## Change of paradigm

Active skills are the `list of last used skills ordered by timestamp`, the 5 minutes timeout is a failsafe, if a user walks away from an interaction it forcefully deactivates the skill

I appreciate the intention of the state management, but the state of active/inactive is not up to skill to decide, thats a skills_service function.

There are helper methods individual skills can use to activate themselves, or that external services can use to activate a skill (eg, the GUI). There is this old PR where i tried to [add a disable callback](https://github.com/MycroftAI/mycroft-core/pull/1468), i think it's a good idea to also add a method for skills to deactivate themselves early

## other considerations

Any skill can make itself a "system" skill and capture all utterances, its not a robust design

## Breaking change

This is also a breaking change, it changes the definition of what an active skill is, there is an [official api](https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/skills/intent_service_interface.py#L251) to retrieve this and there are [downstream projects](https://github.com/OpenVoiceOS/ovos_utils/blob/dev/ovos_utils/intents/__init__.py) that depend on this definition

## Alternative approaches

I think the enabling/disabling of intents is a great idea, and it makes a lot of sense that skills should also be able to manage their own state separate from the active skill paradigm

I have an implementation of `IntentLayers` exactly for this use case, i updated it after i saw this PR and you can find it [here](https://github.com/OpenVoiceOS/OVOS-workshop/blob/master/ovos_workshop/skills/layers.py) and [decorators here](https://github.com/OpenVoiceOS/OVOS-workshop/blob/master/ovos_workshop/skills/decorators/layers.py), updated demo skill coming soon

forslund commented on mycroft/skills/intent_service.py:
```diff
@@ -296,14 +321,30 @@ def handle_utterance(self, message):
+                        """
+                        # stop last active common play skill
+                        msg_data = {"stop_type":"latest"}
+                        self.bus.emit(message.forward('play:stop', msg_data))
```
This seems really weird, can you elaborate of what the intention is to stop any playing media for "undefined" skills. I don't have access to the Jira ticket(s) this PR refers to.

Perhaps it's a good idea to break this out into a named method with a docstring explaining what it does and how it ties into the rest of the system.

ken-mycroft replied:
Ah, good catch Ake. So this is an interesting change which was not present in the previous PR similar to this new one. I believe it was called SkillStates. This change is a direct result of enabling barge-in which was necessary to be able to stop long playing dialog skills like wiki and the definition of some basic system level interactions.

When looking for UI guidance on how skills should interact it was stated that if a skill is playing (say for example a music skill) and another skill gets launched (say while the music is playing I barge-in and say 'tell me about the Beatles') then the old skill should be stopped (music) and the new skill should take over (wiki). It should also work the other way around. If wiki were playing and I said 'play louie louie' the music skill activation should cause the wiki skill to stop. However, if an alarm goes off while the music is playing and I say stop, the alarm should stop but not the music skill. Likewise, while I am playing music, if I say 'set the volume to 50%' that should also, not stop the music skill. The ability to handle different more complex scenarios like this obviously led to the skill category I added in this design, however, I still needed to make some trade-offs. 

For example, the code in question works if the underlying skill is a newer style skill like a common play skill or a statefull skill, however, if the underlying skill is an old style skill (normal skill as they exist today) then they will work as always which is probably not what we will want so that problem still has to be solved. I felt even though this would change default behavior it was halfway to the ultimate solution. 

This is a contentious change and unlike anything else in this PR could be considered a change in existing behavior. I would not be adverse to simply commenting out these two lines for now and leaving the default skill behavior unchanged.

JarbasAl replied:
can you make this behavior configurable? a `"common_play_barge_in": true` in the .conf file maybe? 
we could either add a whole section for `"common_play"` assuming more config options will be needed in the future, or a simple boolean under the "skills" section

ken-mycroft replied:
This is an interesting 'ask'. So barge-in is currently controlled by a value in the default config file 

"mute_during_output" : true,

Gez has requested in a different comment on this PR to remove the change (I changed it to false) from here and put it into the system conf file. Do you think I should use this value in the decision making process or maybe even should it be configurable at a skill level? Be interested in hearing Gez's thoughts on this as well.

JarbasAl replied:
here is the scenario im thinking about that makes me think this should be optional in the first place (the mic quality is also a consideration)

- music is playing
- microphone is good, hears wakeword, lowers volume, receives utterance, restores volume
- music is playing like nothing happened
- my intent was a simple non intrusive command

the concern is that this new stop signal makes music stop completely a lot of the time

ken-mycroft replied:
Ah, so yes, you are correct and this was a conscience decision though I could be talked out of it but let me explain. 

The ask from our UI department was that if you were playing a song and I said 'tell me about bla' that the song would stop and the new skill would take over. This was not the desired behavior, however, in the case of a system type event like an alarm or a request to increase the volume. In this case they should not terminate the underlying active skill. 

So, this bit of code does just that. If you look at the alarm and/or the volume skill you will see they set their category to 'system' which basically means they will not stop an underlying skill, but older skills like wiki or ddg will indeed stop the underlying skill. It was a tough decision because this really only solves half the problem. When an underlying skill is an old style skill it will behave as they do now and not stop when another skill gains control of the output channel. I would like to solve this also but have not decided how to do that yet. 

So do you think I should remove these two lines of code until the entire (both halves) problem has been solved or do you think I should leave this here and allow skills to change their category to deal with this. Keep in mind only category = system is considered currently but that could be easily changed.

ken-mycroft replied:
And let me add if you check out the new version of the volume skill in conjunction with this code it will behave as you mention. It will change the volume and resume the media play. Similarly the alarm skill will respond to stop and not take out any underlying skill; new style or old style. It will just stop the alarm.

forslund commented:
Since I was asked to review I'd like to know more about what the intention of these changes are. To me they look like 3 or 4 different things all in 1 commit with basically no information. If possible split it up into more logical commits so it's easier to understand which parts are related and which are separate changes.

The skill state thing I understand I think but the reasons for the changes in the converse and how it would work is particularly cloudy to me.

ken-mycroft commented on mycroft/skills/intent_service.py:
```diff
@@ -17,13 +17,15 @@
 import time
 
 from mycroft.configuration import Configuration, set_default_lf_lang
+from mycroft.util.lang import set_active_lang
```
This is actually used in reset converse() so if I remove it it might(?) break something. That pre-dates me and I have no idea about that code. Am I to understand that the feature/mark-2 branch is broken? I branched off feature/mark-2 and I did not change this code?

ken-mycroft replied:
Oh I see . this is because this PR is like 2 weeks old now and the underlying code has changed recently. OK, so what is it supposed to be. Is this what replaces set_active_lang?

set_default_lf_lang(lang)

ken-mycroft replied:
OK, I believe I have taken care of this

ken-mycroft commented on mycroft/configuration/mycroft.conf:
```diff
@@ -178,7 +178,7 @@
     // This is enabled by default, but instances with good microphone noise cancellation
     // can disable this to listen all the time, allowing 'barge in' functionality.
-    "mute_during_output" : true,
+    "mute_during_output" : false,
```
Agreed, however, this PR has taken on a life of its own and I am loathe to make any changes yet until I get all these comments resolved so if per chance I forget to revert this, please remind me when the final changes to this PR are made.

ken-mycroft replied:
OK, I took care of this in the branch but not sure how to set it otherwise.

pep8speaks commented:
Hello @ken-mycroft! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`mycroft/skills/common_play_skill.py`](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/common_play_skill.py):

> [Line 173:54](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/common_play_skill.py#L173): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 174:13](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/common_play_skill.py#L174): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 174:52](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/common_play_skill.py#L174): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 174:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/common_play_skill.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

* In the file [`mycroft/skills/intent_service.py`](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py):

> [Line 120:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L120): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 79 characters)
> [Line 158:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L158): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 188:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L188): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)
> [Line 243:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L243): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 296:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L296): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 79 characters)
> [Line 327:73](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L327): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 328:75](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L328): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 330:42](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L330): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 333:48](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L333): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 336:62](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L336): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 337:33](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L337): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 339:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L339): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 373:9](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/intent_service.py#L373): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

* In the file [`mycroft/skills/mycroft_skill/mycroft_skill.py`](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py):

> [Line 172:44](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L172): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 173:45](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L173): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 175:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L175): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 79 characters)
> [Line 181:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L181): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 197:21](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L197): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 202:21](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L202): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 202:60](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L202): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 202:80](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L202): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 712:8](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L712): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 717:9](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/mycroft_skill.py#L717): [E113](https://duckduckgo.com/?q=pep8%20E113) unexpected indentation

* In the file [`mycroft/skills/mycroft_skill/skill_control.py`](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py):

> [Line 17:42](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L17): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 21:43](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L21): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 23:38](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L23): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 24:36](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L24): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 26:40](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L26): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 34:43](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L34): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 35:42](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L35): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 40:43](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L40): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 49:36](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L49): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 55:1](https://github.com/MycroftAI/mycroft-core/blob/ccd41717d2b635eec62adbfd716eb8853afdebe9/mycroft/skills/mycroft_skill/skill_control.py#L55): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

ken-mycroft commented:
FYI - Thanks everybody for their input. I believe the latest commit addresses most of the issues brought up. The biggest outstanding issue I see is the new behavior which stops active common play skills by default for old style skills. If so desired I can roll this change back pretty easily by commenting out a single line. I still have to change the %s to formats{} for Chris V but other than that I think this ready to test. There was also a recommendation that barge-in be configurable and used almost at a skill level and I believe this idea deserves further attention in the near future. For now it is left to Gez to override this configuration setting which will affect behavior at the system level.

chrisveilleux commented:
There appears to be one open conversation left.  I couldn't tell if that was considered "resolved" or not.

NeonDaniel commented:
> I don't agree with this PR implementation, i think it misunderstands the way the converse system works as illustrated by the #TODO fix bug comment.
> 
> > > > > First let me state you are correct in that I do not understand the need for converse. My concerns with this method is that skills seemed to be relying on this method to figure out where they were in their life cycle as well as trying to understand what was going on at a system level. To me converse is simply a method which was used to handle the fact that the skills did not overtly maintain their state. It also seems to be used to handle the difficult issues surrounding propagation of 'stop' intents.
> 
> If you are referring to the 5 minute timeout comment I think I do understand what is going on, however, an arbitrary 5 minute inactivity delay does not help me keep the active_skills array in sync with reality. One of my overarching goals of this PR is to make the active_skills array useful at a higher level. This is why I added the inactivate messages and associated processing, so skills can communicate when they go inactive and the change can be reflected in the active_skills array almost immediately instead of waiting around for 5 minutes. If a skill has gone inactive why should it still live in the active skills array?
> 
> ## Change of paradigm
> Active skills are the `list of last used skills ordered by timestamp`, the 5 minutes timeout is a failsafe, if a user walks away from an interaction it forcefully deactivates the skill
> 
> I appreciate the intention of the state management, but the state of active/inactive is not up to skill to decide, thats a skills_service function.
> 
> > > > > > > Why do you believe a skill's state is not up to the skill? Is the new skill base class 'change_state()' method not considered a part of the skill service function? Perhaps I am missing something here. Please explain.
> 
> ## Breaking change
> This is also a breaking change, it changes the definition of what an active skill is, there is an [official api](https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/skills/intent_service_interface.py#L251) to retrieve this and there are [downstream projects](https://github.com/OpenVoiceOS/ovos_utils/blob/dev/ovos_utils/intents/__init__.py) that depend on this definition
> 
> > > > > > These changes do not change anything to do with the active_skills array except how quickly inactive skills are removed from it (used to be 5 minutes, now it is immediate). The existing messages still work as they always have. The structure of an entry simply added an additional value to the end of a list. It should not break anything. Am I missing something here? Ake actually recommended a named tuple here but that would probably be a breaking change so I am glad you pointed that out but I am confused why you say this will break something. Please elaborate, as it is quite possible I may have missed something.
> 
> ## Alternative approaches
> I think the enabling/disabling of intents is a great idea, and it makes a lot of sense that skills should also be able to manage their own state separate from the active skill paradigm
> 
> I have an implementation of `IntentLayers` exactly for this use case, i updated it after i saw this PR and you can find it [here](https://github.com/OpenVoiceOS/OVOS-workshop/blob/master/ovos_workshop/skills/layers.py) and [decorators here](https://github.com/OpenVoiceOS/OVOS-workshop/blob/master/ovos_workshop/skills/decorators/layers.py), updated demo skill coming soon
> 
> > > > > > > I appreciate this, however, this PR fixes the basic issues regarding how skills interact and is backwards compatible with our existing system and does not cause any breaking changes so that was one of my primary goals. There are many potential alternative approaches to solving this issue but there are also only so many hours in a day.
> 
> There are helper methods individual skills can use to activate themselves, or that external services can use to activate a skill (eg, the GUI). There is this old PR where i tried to [add a disable callback](https://github.com/MycroftAI/mycroft-core/pull/1468), i think it's a good idea to also add a method for skills to deactivate themselves early
> 
> ## other considerations
> Any skill can make itself a "system" skill and capture all utterances, its not a robust design
> 
> > > > > > This is a valid concern and was previously raised by Gez. I suppose I could create a white list at the system level somewhere which enumerates the 'system' level skills by id to head off any non-well behaved skill developers. My objective was to introduce a concept which could be leveraged by skill developers such that if they were creating a skill which could be considered a system level skill they could leverage this value as well and benefit from the system level behaviors supported by this category of skill which will hopefully become convention.
> 
> I think you are conflating two different concepts here; one is categorization of skills and the second is capturing of all utterances. Tagging a skill as category 'system' simply gives it priority during converse intent processing which ultimately leads to an ability to terminate the 'stop' intent processing at a certain level in the skill stack. I am pretty sure with our existing design any skill could gobble up all utterances anyway, this change actually circumvents that if a system skill is involved in the process.
> 
> I am actually conflicted whether this should be something that is extended to other skill developers or whether this is an internal only value. I obviously understand the concern with potential bad actors.

Just wanted to add my thoughts/current understanding of skill structure/lifecycle.

RE Converse: My understanding of the converse system is that it is intended to allow a skill to interact with arbitrary inputs from a user, given there was a recently matched intent. From a utility standpoint, I think this makes sense to more easily capture user inputs without an intent match ('i.e.: capturing generic confirmations or answers to other questions the skill asks in dialog).

RE Active skills: I agree that the arbitrary 5 minute timeout is imperfect, and in NeonCore we changed this to be a configured timeout in `make_active`, so a skill could decide it is active for 0 minutes, or indefinitely, or any time in between. A `make_inactive` method could allow a skill to remove itself from the active list (looks like that is here), though being there only means a converse method is called, and the skill can simply return if it decides it isn't active.

RE Jarbas' change of paradigm comment: I see utility in only allowing a skill to remain active for 5 minutes to prevent unexpected behavior (like an active skill consuming all inputs forever). I think there should be some method for a skill to remain active, given some conditions. For example, we have a skill where a user runs a scripted interaction and they might want that to persist even if they walk away (https://github.com/NeonGeckoCom/custom-conversation.neon).

RE Breaking changes: The skill I linked above is one example of a would-be broken under Mycroft Core skill. I'm not sure of other community skills out there, but this PR does change documented/currently expected behavior.

chrisveilleux reviewed (changes requested):
For a change this fundamental, there should be some unit tests.

JarbasAl commented:
@ken-mycroft please do not edit previous comments to add responses:
- readers see things i did not write under my avatar
- its unclear who is saying what
- wont even notice if you change what i originally said
- changes formatting and order of what i wrote, and therefore messes up with what im trying to convey
- github notifications goes crazy, i have no clue what changed, only that it was in this PR
- you can mess up and accidentally delete whole pieces of the convo
- if you are evil you can even delete it from commit history
- multiple edits from multiple people make readers brains hurt

i have restored my original comment back to its original form, now that thats out of the way....

you said
> I believe ultimately even in converse an intent is matched. I could be wrong. This PR does not change how converse works with the exception that now skills categorized as 'system' will get a shot at handling the utterances while in converse mode, first.

### Purpose of converse method:
- allow skills to intercept an utterance before the intent stage
- allow skills to parse utterances without using the intent parsers
- allow skills to prepare for follow up intent parsing (manage internal state)

converse methods **does not** use regular intent parsing


### Active skill list:
- core manages active skills
  - an active skill is a `skill that returned True in converse method or triggered an intent`
  - the active skill list is ordered by activation timestamp
  - a skill is deactivated if "activate" is not called for 5 minutes (user abandoned interaction)
- skills can request to be activated
- skills can request to be deactivated (missing feature, added by this PR)
- skills should be able to react to being activated/deactivated (missing feature see https://github.com/MycroftAI/mycroft-core/pull/1468)
- there is an official messagebus interface to retrieve this info, external projects depend on it
- there should be limitations in how many times / if a skill can activate itself (missing feature)

### The breaking change explained

- skills now decide if they are active, instead of core itself deciding
- skills that have a category might not even make it to the active skill list
- for skill developers this subtly changes behaviour of `self.make_active` and makes it depend on unrelated settings (`self.category`)
- active skills list can no longer be depended to mean something useful, tells you a skill is there, but there are no constraints, maybe the skill developer felt like activating it every second, maybe the skill isnt there at all because its "new style"
- any downstream using the api to retrieve active skill list is getting a `different thing`

### The confusion

i think the enabling/disabling intents is good, but it has absolutely nothing to do with converse method, so this PR is doing a few different things:

- messing up with converse and active list meaning/execution <- the contentious change
   - adding the deactivate method should be kept
- adding a stop message for common play skills <- should be a different PR
- adding the equivalent of native IntentLayers integration i linked above (why reinvent the wheel?) <- should be a different PR

ken-mycroft commented:
As a result of community feedback it was decided to revert the change that automatically sends out a common play stop message when a new skill is added to the active skills array. Existing skills should now see no difference in their existing behavior.

ken-mycroft commented:
And there was much rejoicing :-)
